From: srinivas pulukuru <srinivas.pulukuru@ti.com>
Date: Tue, 19 Apr 2011 06:51:39 +0000 (-0500)
Subject: OMAP4 DSS HDMI: Ignore HPD and timing change event when connect to HDMI
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~27
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=91ffeccec0cc3d55ca5c6cfbe47766ba82a10090

OMAP4 DSS HDMI: Ignore HPD and timing change event when connect to HDMI

When overlays are still attached to the HDMI panel then no timing
change should be allowed.
Test Equipments generate HPD signals before proceeding to next testcase and
this causes usrepace overlay misconfiguration due to timing change when
video playback is still going on. Ignore HPD events when overlays are
still attached to the HDMI and reset IRQ's before returning.
Also ignore any forced timing change when overlays are connected to
HDMI.

Change-Id: Ibc3cd15d8bd29c6a6eab63ee903ec0359247fbf2
Signed-off-by: srinivas pulukuru <srinivas.pulukuru@ti.com>
---

diff --git a/drivers/video/omap2/dss/hdmi.c b/drivers/video/omap2/dss/hdmi.c
index 32f880e..f45656b 100644
--- a/drivers/video/omap2/dss/hdmi.c
+++ b/drivers/video/omap2/dss/hdmi.c
@@ -1507,12 +1507,35 @@ done:
 
 hpd_modify:
 	if (r & HDMI_HPD_MODIFY) {
+		struct omap_overlay *ovl;
+		int i;
+
+		/* check if any overlays are connected to TV and
+		 * return if connected after resetting the IRQ's
+		 */
+		for (i = 0; i < dssdev->manager->num_overlays; i++) {
+			ovl = dssdev->manager->overlays[i];
+			if (!(strcmp(ovl->manager->name, "tv")))
+				if (ovl->info.enabled) {
+					DSSINFO("Overlay %d is still "
+						"attached to tv\n", ovl->id);
+					DSSINFO("Cannot Rconfigure HDMI when "
+						"overlays are still attached "
+						"to tv\n"
+						"Dettach the overlays before "
+						"reconfiguring the HDMI\n\n");
+					/* clear the IRQ's*/
+					hdmi_set_irqs(0);
+					goto done2;
+				}
+		}
+
 		/*
 		 * HDMI HPD state low followed by a HPD state high
 		 * with more than 100ms duration is recieved so EDID should
 		 * reread and HDMI reconfigued.
 		 */
-		DSSINFO("Reconfigure HDMI PHY - HDMI_HPD_MODIFY\n");
+		DSSINFO("Reconfigure HDMI PHY - HDMI_HPD_MODIFY\n\n");
 		/* force a new power-up to read EDID */
 		edid_set = false;
 		custom_set = false;
@@ -1522,6 +1545,7 @@ hpd_modify:
 		DSSINFO("Reconfigure HDMI PHY Done- HDMI_HPD_MODIFY\n\n");
 	}
 
+done2:
 	mutex_unlock(&hdmi.lock_aux);
 	mutex_unlock(&hdmi.lock);
 	kfree(work);
@@ -1917,10 +1941,28 @@ static void hdmi_set_timings(struct omap_dss_device *dssdev,
 static void hdmi_set_custom_edid_timing_code(struct omap_dss_device *dssdev,
 							int code, int mode)
 {
+	struct omap_overlay *ovl;
+	int i;
+
 	/* for now only set this while on or on HPD */
 	if (dssdev->state != OMAP_DSS_DISPLAY_ACTIVE)
 		return;
 
+	/* check if any overlays are connected to TV and return if connected*/
+	for (i = 0; i < dssdev->manager->num_overlays; i++) {
+		ovl = dssdev->manager->overlays[i];
+		if (!(strcmp(ovl->manager->name, "tv")))
+			if (ovl->info.enabled) {
+				DSSINFO("Overlay[%d] is still "
+					"attached to tv\n", ovl->id);
+				DSSINFO("Cannot change HDMI timing when "
+					"overlays are still attached to tv\n"
+					"Dettach the overlays before "
+					"timing change\n\n");
+				return;
+			}
+	}
+
 	/* ignore invalid codes */
 	if ((!mode && code < ARRAY_SIZE(code_vesa) && code_vesa[code] >= 0) ||
 	    (mode == 1 && code < ARRAY_SIZE(code_cea) && code_cea[code] >= 0)) {